src/dbnode/persist/fs/merger.go

```diff
@@ -0,0 +1,255 @@
+
+	// First stage: loop through series on disk.
+	for id, tagsIter, data, _, err := reader.Read(); err != io.EOF; id, tagsIter, data, _, err = reader.Read() {
+		if err != nil {
+			multiErr = multiErr.Add(err)
+			return multiErr.FinalError()
+		}
+
+		// Reset BlockReaders.
+		brs = brs[:0]
+		br = br[:0]
+		br = append(br, blockReaderFromData(data, segReader, startTime, blockSize))
+
+		// Check if this series is in memory (and thus requires merging).
+		encoded, hasData, err := mergeWith.Read(id, blockStart, nsCtx)
+		if err != nil {
+			multiErr = multiErr.Add(err)
+			return multiErr.FinalError()
+		}
+		if hasData {
+			br = append(br, encoded...)
+		}
+		brs = append(brs, br)
+
+		multiIter.ResetSliceOfSlices(xio.NewReaderSliceOfSlicesFromBlockReadersIterator(brs), nsCtx.Schema)
+
+		// tagsIter is never nil.
+		tags, err := convert.TagsFromTagsIter(id, tagsIter, identPool)
+		tagsIter.Close()
+		if err != nil {
+			multiErr = multiErr.Add(err)
+			return multiErr.FinalError()
+		}
+		if err := persistIter(prepared.Persist, multiIter, id, tags, encoderPool); err != nil {
+			multiErr = multiErr.Add(err)
+			return multiErr.FinalError()
+		}
+	}
+	// Second stage: loop through rest of the merge target that was not captured
+	// in the first stage.
+	err = mergeWith.ForEachRemaining(blockStart, func(seriesID ident.ID, tags ident.Tags) bool {
+		encoded, hasData, err := mergeWith.Read(seriesID, blockStart, nsCtx)
+		if err != nil {
+			multiErr = multiErr.Add(err)
+			return false
+		}
+
+		if hasData {
+			multiIter.Reset(xio.NewSegmentReaderSliceFromBlockReaderSlice(encoded), startTime, blockSize, nsCtx.Schema)
+			if err := persistIter(prepared.Persist, multiIter, seriesID, tags, encoderPool); err != nil {
+				multiErr = multiErr.Add(err)
+				return false
+			}
+		}
+
+		return true
+	})
+	if err != nil {
+		multiErr = multiErr.Add(err)
+		return multiErr.FinalError()
+	}
+
+	// Close the flush preparer, which writes the rest of the files in the
+	// fileset.
+	if err := prepared.Close(); err != nil {
+		multiErr = multiErr.Add(err)
+		return multiErr.FinalError()
+	}
+
+	return multiErr.FinalError()
+}
+
+func blockReaderFromData(
+	data checked.Bytes,
+	segReader xio.SegmentReader,
+	startTime time.Time,
+	blockSize time.Duration,
+) xio.BlockReader {
+	seg := ts.NewSegment(data, nil, ts.FinalizeHead)
+	segReader.Reset(seg)
+	return xio.BlockReader{
+		SegmentReader: segReader,
+		Start:         startTime,
+		BlockSize:     blockSize,
+	}
+}
+
+func persistIter(
+	persistFn persist.DataFn,
+	it encoding.Iterator,
+	id ident.ID,
+	tags ident.Tags,
+	encoderPool encoding.EncoderPool,
+) error {
+	encoder := encoderPool.Get()
+	for it.Next() {
+		if err := encoder.Encode(it.Current()); err != nil {
+			encoder.Close()
+			return err
+		}
+	}
+	if err := it.Err(); err != nil {
+		return err
+	}
+
+	segment := encoder.Discard()
+	checksum := digest.SegmentChecksum(segment)
+
+	err := persistFn(id, tags, segment, checksum)
+	id.Finalize()
```

Comment:
Thanks for the catch. Having disk IDs finalizable but not for the IDs from memory makes this logic difficult to integrate with at the writer layer, since you'd need two different flush preparers with different open options.  I'll just keep track of the IDs from disk and finalize them at the end.